### PR TITLE
catalog: Add new RoleId Variant for group roles

### PR DESCRIPTION
**Pull request by jkosh44**
Previously, the `RoleId::System` was used for all roles that were built in to the system. Generally these roles fall into one of two categories:

  1. IDs that are used by the system or by the support team to log into a Materialize instance and perform some action. These IDs are not meant to be touched by or used by any user. Currently, these are the `mz_system` and `mz_support` roles.
  2. IDs that are built into the system that are granted to users to provide some specific privilege. These IDs are meant to be exclusively used by users and not used by the system. Currently, these are the `mz_monitor` and `mz_monitor_redacted` roles.

These two role types often have very differently behavior and require us to dynamically check which system role type a specific system role is.

This commit adds a new `RoleId` variant for builtin roles of category (2) called `Group`. Builtin roles of category (1) will continue to use the existing `System` variant. This will make it much easier to differentiate between the two types of builtin roles.

### Motivation
This PR refactors existing code.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - There are no user-facing behavior changes.

### CODE REVIEW

**Comment by jkosh44** [comment]
Bike shedding on the `Group` name is welcome.

**Comment by jkosh44** [comment]
Also feel free to push back against this PR completely. It makes RBAC checks much easier since System roles and Group roles have fundamentally different behaviors in RBAC.

File: src/catalog/src/durable/upgrade/v53_to_v54.rs [hunk 1]
```diff
@@ -0,0 +1,441 @@
...
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use mz_stash::upgrade::WireCompatible;
+use mz_stash::wire_compatible;
+
+use crate::durable::upgrade::MigrationAction;
+use crate::durable::upgrade::{objects_v52 as v52, objects_v54 as v54};
```
Comment (ParkMyCar) [inline src/catalog/src/durable/upgrade/v53_to_v54.rs:14]: shouldn't these be v53, not v52?

**Review by ParkMyCar — COMMENTED** [review]
Overall I like this idea, but I don't think `Group` is the right term.

I also thought about calling it `Admin` or `Superuser`, but those don't seem right either? And possibly confusing since they would overlap with Frontegg auth.

It seems like these are roles that are related to the introspection or administration of the system in some way so I'm kind of leaning towards something like `SysAdmin`, but don't love it either

File: src/catalog/src/durable/upgrade/v53_to_v54.rs (hunk 1 above)
Comment (jkosh44) [reply src/catalog/src/durable/upgrade/v53_to_v54.rs:14]: Eeek, yes, good catch. Looks like I missed this when resolving conflicts.

Comment (jkosh44) [reply src/catalog/src/durable/upgrade/v53_to_v54.rs]: Fixed

**Comment by jkosh44** [comment]
> Overall I like this idea, but I don't think `Group` is the right term.

I have the same feeling.

> I also thought about calling it `Admin` or `Superuser`, but those don't seem right either? And possibly confusing since they would overlap with Frontegg auth.

My issue with these is that the name doesn't really differentiate them from the `System` roles. 

> It seems like these are roles that are related to the introspection or administration of the system in some way so I'm kind of leaning towards something like `SysAdmin`, but don't love it either

They're not specifically related to administration, they are just associated with a certain privilege in the system. Specifically reading from unredacted system tables in this case. Some other ideas I had that I didn't really like are:

- `Grantable`
- `SystemGrantable`
- `Privileged`
- `Builtin`

**Comment by ParkMyCar** [comment]
I kind of like `Builtin`? Courtesy of ChatGPT what do you think of these names:
* `Embedded`
* `Intrinsic`
* `Audit`
* `Observe`
* `Control`

**Comment by jkosh44** [comment]
I'm starting to have second thoughts about this PR. Originally the `mz_monitor` and `mz_monitor_redacted` roles were pitched as temporary hacks. So maybe this PR is solidifying their design forever, which is not great. On the other hand, they already exist and I don't think anyone is going to change them anytime soon so we may as well improve their ergonomics.

**Review by mjibson — APPROVED** [review]
Why were these supposed to be temporary? Were they workarounds for some larger problem? Should we decide they are good and not temporary?

**Comment by jkosh44** [comment]
> Why were these supposed to be temporary? Were they workarounds for some larger problem? 

We wanted to be able to grant and revoke `SELECT` privileges from certain builtin objects that expose statement logging information. However, system object privileges are compiled into the binary and cannot be changed. So we made these special system roles that always have certain privileges on certain builtin objects. Those system roles are then granted to users.

> Should we decide they are good and not temporary?

I have a feeling that no matter what they decide they'll probably be around for awhile.

File: src/adapter/src/catalog/state.rs [hunk 2]
```diff
@@ -2156,8 +2155,7 @@ impl CatalogState {
     }
 
     pub fn ensure_grantable_role(&self, role_id: &RoleId) -> Result<(), Error> {
-        let is_grantable = !role_id.is_public()
-            && ((!role_id.is_system()) || GRANTABLE_BUILTIN_ROLE_IDS.contains(role_id));
+        let is_grantable = !role_id.is_public() && !role_id.is_system();
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply src/adapter/src/catalog/state.rs:2158]: Yeah, `mz_monitor` and `mz_monitor_redacted` are grantable to users.

File: src/catalog/src/durable/transaction.rs [hunk 3]
```diff
@@ -280,8 +280,8 @@ impl<'a> Transaction<'a> {
         oid: u32,
     ) -> Result<RoleId, CatalogError> {
         soft_assert_or_log!(
-            id.is_system() || id.is_public(),
-            "ID {id:?} is not system or public variant"
+            id.is_system() || id.is_public() || id.is_group(),
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply src/catalog/src/durable/transaction.rs]: Done

File: src/adapter/src/catalog/state.rs [hunk 4]
```diff
@@ -2179,6 +2177,17 @@ impl CatalogState {
         }
     }
 
+    pub fn ensure_not_group_role(&self, role_id: &RoleId) -> Result<(), Error> {
+        if role_id.is_group() {
+            let role = self.get_role(role_id);
+            Err(Error::new(ErrorKind::ReservedSystemRoleName(
+                role.name().to_string(),
+            )))
```
Comment (nrainer-materialize) [inline src/adapter/src/catalog/state.rs:2185]: Not tested according to coverage information

File: src/catalog/src/durable/upgrade.rs [hunk 5]
```diff
@@ -252,52 +253,50 @@ pub(crate) async fn upgrade(
...
-            51 => {
-                run_versioned_upgrade(unopened_catalog_state, mode, version, v51_to_v52::upgrade)
-                    .await
-            }
-            52 => {
-                run_versioned_upgrade(unopened_catalog_state, mode, version, v52_to_v53::upgrade)
-                    .await
-            }
+    match version {
+        ..=TOO_OLD_VERSION => Err(incompatible),
```
Comment (nrainer-materialize) [inline src/catalog/src/durable/upgrade.rs:275]: Not tested according to coverage information (based on the tests pipeline). Is this plausible? Though, I think it should be tested by the upgrade tests in Nightly, right?

**Review by nrainer-materialize — COMMENTED** [review]
I triggered a [nightly](https://buildkite.com/materialize/nightlies/builds/7224) and a [coverage](https://buildkite.com/materialize/coverage/builds/403) build. The nightly build looks good.

File: src/catalog/src/durable/upgrade.rs (hunk 5 above)
Comment (jkosh44) [reply src/catalog/src/durable/upgrade.rs:275]: This error should be unreachable except for programmer error. 

In order to test this code we would need to start Materialize on some version `x` s.t. the catalog version is `v`. Then directly upgrade materialize to version `y` s.t. the migration from catalog version `v` is removed. However, the check for skipping Materialize versions would be triggered before this check.

**Comment by jkosh44** [comment]
I think I have the perfect name, `Predefined`. It matches the definition of `Predefined` that PostgreSQL uses: https://www.postgresql.org/docs/current/predefined-roles.html

File: src/catalog/protos/objects.proto [hunk 6]
```diff
@@ -694,7 +695,7 @@ message AuditLogEventV1 {
     RenameSchemaV1 rename_schema_v1 = 27;
     UpdateItemV1 update_item_v1 = 26;
     CreateSourceSinkV3 create_source_sink_v3 = 29;
-        AlterRetainHistoryV1 alter_retain_history_v1 = 30;
+    AlterRetainHistoryV1 alter_retain_history_v1 = 30;
```
Comment (jkosh44) [inline src/catalog/protos/objects.proto:698]: I guess we don't lint our .proto files? I'm tempted to just leave this all in because why not.

**Review by mjibson — APPROVED** [review]

**Pull request merged**